cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+/* file: tsne_gradient_descent_avx512_impl.i */
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Parts of tSNE algorithm oprtimized for AVX512.
+//--
+*/
+
+#ifndef __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+#define __TSNE_GRADIENT_DESCENT_AVX512_IMPL_I__
+
+namespace daal
+{
+namespace algorithms
+{
+namespace internal
+{
+
+/* Partial template specialization of attractive kernel for single precision data and AVX512 ISA */
+template <bool DivComp, typename IdxType>
+struct AttractiveKernel<DivComp, IdxType, float, avx512>
+{
+    static services::Status impl(const float * val, const IdxType * col, const size_t * row, MemoryCtxType<IdxType, float, avx512> & mem,
+                                 float & zNorm, float & divergence, const IdxType N, const IdxType nnz, const IdxType nElements,
+                                 const float exaggeration)
+    {
+        const float multiplier = exaggeration * float(zNorm);
+        divergence             = 0.;
+
+        const IdxType prefetch_dist = 32;
+
+        daal::TlsSum<float, avx512> divTlsData(1);
+
+        SafeStatus safeStat;
+        daal::tls<float *> logTlsData([=, &safeStat]() {
+            auto logData = services::internal::service_scalable_calloc<float, avx512>(nElements);
+            if (!logData)
+            {
+                safeStat.add(services::ErrorMemoryAllocationFailed);
+            }
+            return logData;
+        });
+
+        const IdxType nThreads    = threader_get_threads_number();
+        const IdxType sizeOfBlock = services::internal::min<avx512, size_t>(256, N / nThreads + 1);
+        const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
```

Comment:
1. Please use explicit bool conversion instead
```suggestion
        const IdxType nBlocks     = N / sizeOfBlock + bool(N % sizeOfBlock);
```
2. Please check that `nBlocks * sizeOfBlock >= N`